loader/src/references/reference_binder.py

```diff
@@ -0,0 +1,137 @@
+        for reference_path in reference_paths:
+            logger.debug(
+                f"Trying to resolve reference for resource {fhir_object['id']}"
+                f"at {reference_path}"
+            )
+            try:
+                bound_ref = self.bind_existing_reference(fhir_object, reference_path)
+                fhir_object[reference_path] = bound_ref
+            except Exception as e:
+                logger.warning(
+                    "Error while binding reference for instance "
+                    f"{fhir_object} at path {reference_path}: {e}"
+                )
+
+        if fhir_object.get("identifier") and len(fhir_object["identifier"]) > 0:
+            self.resolve_pending_references(fhir_object)
+
+        return fhir_object.to_dict()
+
+    def bind_existing_reference(self, fhir_object, reference_path):
+        # FIXME: dotty-dict does not handle brackets indices,
+        # it uses dots instead (a.0.b instead of a[0].b)
+        reference_attribute = fhir_object[reference_path]
+
+        # If we have a list of references, we want to bind all of them.
+        # Thus, we loop on all the items in sub_fhir_object.
+        if not isinstance(reference_attribute, list):
+            reference_attribute = [reference_attribute]
+
+        for ref in reference_attribute:
+            # extract the type and itentifier of the reference
+            reference_type = ref["type"]
+            identifier = ref["identifier"]
+            if not identifier.get("value") or not identifier.get("system"):
+                logger.error(
+                    f"invalid reference: {ref}. identifier.value and identifier.system are required"
+                )
+                continue
+
+            # search the referenced resource in the database
+            referenced_resource = self.fhirstore.db[reference_type].find_one(
+                {"identifier": {"value": identifier["value"], " system": identifier["system"]}},
+                ["id"],
+            )
+            if referenced_resource:
+                # if found, add the ID as the "literal reference"
+                # (https://www.hl7.org/fhir/references-definitions.html#Reference.reference)
+                logger.info(f"reference to {reference_type} {identifier['value']} resolved")
+                ref["reference"] = f"{reference_type}/{referenced_resource['id']}"
+            else:
+                logger.info(
+                    f"caching reference to {reference_type} "
+                    f"{identifier['value']} at {reference_path}"
+                )
+                # otherwise, cache the reference to resolve it later
+                cache_key = (reference_type, identifier["system"], identifier["value"])
+                pending_refs = self.cache.get(cache_key, {})
+                self.cache[cache_key] = {
+                    **pending_refs,
+                    (fhir_object["resourceType"], reference_path): [
```

Comment:
Could we use a defaultdict(list) here? I think it would be clearer

Reply:
i'll try